azkaban-common/src/main/java/azkaban/executor/ExecutorHealthChecker.java

```diff
@@ -0,0 +1,158 @@
+  // Max number of executor failures before sending out alert emails.
+  private static final int DEFAULT_EXECUTOR_MAX_FAILURE_COUNT = 6;
+  // Web server checks executor health every 5 min by default.
+  private static final Duration DEFAULT_EXECUTOR_HEALTHCHECK_INTERVAL = Duration.ofMinutes(5);
+  private final long healthCheckIntervalMin;
+  private final int executorMaxFailureCount;
+  private final List<String> alertEmails;
+  private final ScheduledExecutorService scheduler;
+  private final ExecutorLoader executorLoader;
+  private final ExecutorApiGateway apiGateway;
+  private final AlerterHolder alerterHolder;
+  private final Map<Integer, Integer> executorFailureCount = new HashMap<>();
+
+  @Inject
+  public ExecutorHealthChecker(final Props azkProps, final ExecutorLoader executorLoader,
+      final ExecutorApiGateway apiGateway, final AlerterHolder alerterHolder) {
+    this.healthCheckIntervalMin = azkProps
+        .getLong(ConfigurationKeys.AZKABAN_EXECUTOR_HEALTHCHECK_INTERVAL_MIN,
+            DEFAULT_EXECUTOR_HEALTHCHECK_INTERVAL.toMinutes());
+    this.executorMaxFailureCount = azkProps.getInt(ConfigurationKeys
+        .AZKABAN_EXECUTOR_MAX_FAILURE_COUNT, DEFAULT_EXECUTOR_MAX_FAILURE_COUNT);
+    this.alertEmails = azkProps.getStringList(ConfigurationKeys.AZKABAN_ADMIN_ALERT_EMAIL);
+    this.scheduler = Executors.newSingleThreadScheduledExecutor();
+    this.executorLoader = executorLoader;
+    this.apiGateway = apiGateway;
+    this.alerterHolder = alerterHolder;
+  }
+
+  public void start() {
+    this.scheduler.scheduleAtFixedRate(() -> checkExecutorHealth(), 0L, this.healthCheckIntervalMin,
+        TimeUnit.SECONDS);
+  }
+
+  public void shutdown() {
+    this.scheduler.shutdown();
+    this.scheduler.shutdownNow();
+  }
+
+  /**
+   * Checks executor health. Finalizes the flow if its executor is already removed from DB or
+   * sends alert emails if the executor isn't alive any more.
+   */
+  public void checkExecutorHealth() {
+    final Map<Optional<Executor>, List<ExecutableFlow>> exFlowMap = getFlowToExecutorMap();
+    for (final Map.Entry<Optional<Executor>, List<ExecutableFlow>> entry : exFlowMap.entrySet()) {
+      final Optional<Executor> executorOption = entry.getKey();
+      if (!executorOption.isPresent()) {
+        final String finalizeReason = "Executor id of this execution doesn't exist.";
+        for (final ExecutableFlow flow : entry.getValue()) {
+          this.logger.warn(
+              String.format("Finalizing execution %s, %s", flow.getExecutionId(), finalizeReason));
+          ExecutionControllerUtils
+              .finalizeFlow(this.executorLoader, this.alerterHolder, flow, finalizeReason, null);
+        }
+        continue;
+      }
+
+      final Executor executor = executorOption.get();
+      try {
+        final Map<String, Object> results = this.apiGateway
+            .callWithExecutionId(executor.getHost(), executor.getPort(),
```

Comment:
Good question. The frequency of this http call is relatively low (5 minutes interval per check), even if each http call takes long, it will still be acceptable in our use case though. For better monitoring, I think we can add some metrics to track the time it takes for each call to return.

Reply:
Metrics sound good, just for general monitoring and debugging purposes as well.